src/xwayland/mod.rs

```diff
@@ -0,0 +1,313 @@
+#![cfg(feature = "xwayland")]
+use std::{cell::RefCell, collections::HashMap, rc::Rc};
+
+use smithay::{
+    desktop::Window,
+    reexports::{calloop::LoopHandle, wayland_server::DisplayHandle, x11rb},
+    utils::{Logical, Rectangle},
+    xwayland::{
+        xwm::{Reorder, ResizeEdge, XwmId},
+        X11Surface, X11Wm, XWayland, XWaylandEvent, XWaylandSource, XwmHandler,
+    },
+};
+#[cfg(feature = "xwayland")]
+use tracing::{debug, error, info};
+
+use crate::{
+    state::{Backend, CalloopData, MagmaState},
+    utils::workspace::MagmaWindow,
+};
+
+#[derive(Debug)]
+pub struct XWaylandState {
+    pub handle: XWayland,
+    pub xwm: Option<X11Wm>,
+    pub xdisplay: Option<u32>,
+}
+
+impl XWaylandState {
+    /// Create a new handle and event source
+    pub fn new(dh: &DisplayHandle) -> (Self, XWaylandSource) {
+        let (handle, source) = XWayland::new(dh);
+
+        (
+            Self {
+                handle,
+                xwm: None,
+                xdisplay: None,
+            },
+            source,
+        )
+    }
+
+    /// Start the xwayland server
+    pub fn start(&mut self, loop_handle: LoopHandle<CalloopData<impl Backend>>) {
+        let env: HashMap<String, String> = HashMap::new();
+
+        self.xdisplay = Some(
+            self.handle
+                .start(loop_handle, None, env, true, |_| {})
+                .expect("Failed to start xwayland server!"),
+        );
+    }
+
+    pub fn on_event<BackendData: Backend>(
+        &mut self,
+        event: XWaylandEvent,
+        loop_handle: LoopHandle<'static, CalloopData<BackendData>>,
+        display_handle: &mut DisplayHandle,
+    ) {
+        match event {
+            XWaylandEvent::Ready {
+                connection,
+                client,
+                client_fd,
+                display,
+            } => {
+                let d = display;
+                std::env::set_var("DISPLAY", format!(":{d}"));
+                info!("Initialized xwayland: fd {}, display {}", client_fd, d);
+                self.xwm = match X11Wm::start_wm(
+                    loop_handle,
+                    display_handle.clone(),
+                    connection,
+                    client,
+                ) {
+                    Ok(wm) => Some(wm),
+                    Err(e) => {
+                        error!(?e, "Failed to start xwaylacnd WM");
+                        None
+                    }
+                };
+            }
+            XWaylandEvent::Exited => {
+                info!("xwayland exited");
+                self.xwm = None;
+                self.xdisplay = None;
+            }
+        }
+    }
+}
+
+impl<BackendData: Backend> XwmHandler for CalloopData<BackendData> {
```

Comment:
you already implemeted XwmHandler for MagmaState why again for calloop data and why are you again falling back to MagmaState anyway im confused